apps/arena/lib/arena/game/skill.ex

```diff
@@ -217,15 +242,94 @@ defmodule Arena.Game.Skill do
 
       Process.send_after(self(), {:remove_effect, player.id, last_id}, effect.duration_ms)
 
-      effects =
-        get_in(game_state, [:players, player.id, :aditional_info, :effects])
-        |> Map.put(last_id, effect)
-
-      put_in(game_state, [:players, player.id, :aditional_info, :effects], effects)
+      put_in(
+        game_state,
+        [:players, player.id, :aditional_info, :effects, last_id],
+        Map.put(effect, :id, last_id)
+      )
       |> put_in([:last_id], last_id)
     end)
   end
 
+  def apply_effect_mechanic(players, game_state) do
+    updated_players =
+      Map.filter(players, fn {_, player} -> Player.alive?(player) end)
+      |> Enum.map(fn {_player_id, player} ->
+        player =
+          Enum.reduce(player.aditional_info.effects, player, fn {_effect_id, effect}, player ->
+            apply_effect_mechanic(player, effect, game_state)
+          end)
+
+        {player.id, player}
+      end)
+      |> Map.new()
+
+    Map.merge(players, updated_players)
+  end
+
+  def apply_effect_mechanic(player, effect, game_state) do
+    now = DateTime.utc_now() |> DateTime.to_unix(:millisecond)
+
+    mechanics_to_apply =
+      effect.effect_mechanics
+      |> Map.filter(fn {_name, mechanic} ->
+        now - Map.get(mechanic, :last_application_time, 0) >= mechanic.effect_delay_ms
+      end)
+      |> Enum.map(fn {name, mechanic} ->
```

Comment:
Same as above, use `Map.new/2`

Reply:
Since i reworked this functions we don't have this map new here